twistrs/src/permutate.rs

```diff
@@ -257,470 +246,325 @@ impl<'a> Domain<'a> {
-        }
-
-        dbg!(&result);
-
-        Domain::filter_domains(Box::new(result.into_iter()))
+    pub fn omission(&self) -> impl Iterator<Item = String> + '_ {
+        Domain::filter_domains(self.fqdn.chars().enumerate().map(move |(i, _)| {
+            let mut permutation = self.fqdn.to_string();
+            permutation.remove(i);
+            permutation
+        }))
     }
 
     /// Permutation method that repeats characters twice provided they are
     /// alphabetic characters (e.g. `google.com` -> `gooogle.com`).
-    pub fn repetition(&self) -> Result<Box<dyn Iterator<Item = String>>> {
-        let mut result: Vec<String> = vec![];
-
-        for (i, c) in self.fqdn.chars().collect::<Vec<char>>().iter().enumerate() {
-            // @CLEANUP(jdb): Any way to do this nicely? Just want to avoid
-            //                out of bounds issues.
-            if i == self.fqdn.len() {
-                break;
-            }
-
+    pub fn repetition(&self) -> impl Iterator<Item = String> + '_ {
+        Domain::filter_domains(self.fqdn.chars().enumerate().filter_map(move |(i, c)| {
             if c.is_alphabetic() {
-                result.push(format!("{}{}{}", &self.fqdn[..=i], c, &self.fqdn[i + 1..]));
+                Some(format!("{}{}{}", &self.fqdn[..=i], c, &self.fqdn[i + 1..]))
+            } else {
+                None
             }
-        }
-
-        dbg!(&result);
-
-        Domain::filter_domains(Box::new(result.into_iter()))
+        }))
     }
 
     /// Permutation method similar to insertion, except that it replaces a given
     /// character with another character in proximity depending on keyboard layout.
-    pub fn replacement(&self) -> Result<Box<dyn Iterator<Item = String>>> {
-        let mut result: Vec<String> = vec![];
-
-        for (i, c) in self.fqdn.chars().collect::<Vec<char>>().iter().enumerate() {
-            // We do not want to insert in the beginning or in the end of the domain
-            if i == 0 || i == self.fqdn.len() - 1 {
-                continue;
-            }
-
-            for keyboard_layout in KEYBOARD_LAYOUTS.iter() {
-                if keyboard_layout.contains_key(c) {
-                    for keyboard_char in keyboard_layout
-                        .get(c)
-                        .unwrap()
-                        .chars()
-                        .collect::<Vec<char>>()
-                        .iter()
-                    {
-                        result.push(format!(
-                            "{}{}{}",
-                            &self.fqdn[..i],
-                            *keyboard_char,
-                            &self.fqdn[i + 1..]
-                        ));
-                    }
-                }
-            }
-        }
-
-        dbg!(&result);
-
-        Domain::filter_domains(Box::new(result.into_iter()))
+    pub fn replacement(&self) -> impl Iterator<Item = String> + '_ {
+        Self::filter_domains(
```

Comment:
Ah, not here specifically, no. Usually I would use `Self::filter_domains` everywhere (I think that's recommended by clippy sometimes? perhaps on a pedantic link), perhaps in case the struct name ever changes - I can't remember why I started tbh!